lib/spack/spack/ci/common.py

```diff
@@ -0,0 +1,826 @@
+    of edge types."""
+
+    @classmethod
+    def key(cls, spec: spack.spec.Spec) -> str:
+        return spec.dag_hash()
+
+    def __init__(self, specs: List[spack.spec.Spec]) -> None:
+        # Build dictionary of nodes
+        self.nodes: Dict[str, PipelineNode] = {
+            PipelineDag.key(s): PipelineNode(s)
+            for s in traverse.traverse_nodes(specs, deptype=dt.ALL_TYPES, root=True)
+        }
+
+        # Create edges
+        for edge in traverse.traverse_edges(
+            specs, deptype=dt.ALL_TYPES, root=False, cover="edges"
+        ):
+            parent_key = PipelineDag.key(edge.parent)
+            child_key = PipelineDag.key(edge.spec)
+
+            self.nodes[parent_key].children.add(child_key)
+            self.nodes[child_key].parents.add(parent_key)
+
+    def prune(self, node_key: str):
+        """Remove a node from the graph, and reconnect its parents and children"""
+        node = self.nodes[node_key]
+        for parent in node.parents:
+            self.nodes[parent].children.remove(node_key)
+            self.nodes[parent].children |= node.children
+        for child in node.children:
+            self.nodes[child].parents.remove(node_key)
+            self.nodes[child].parents |= node.parents
+        del self.nodes[node_key]
+
+    def traverse(self, top_down: bool = True):
+        """If top_down is True, yields the roots first, followed by direct children
+        of roots, followed by their children and so on.  Otherwise yields the leaves
+        first, followed by nodes whose only children were leaves, etc."""
+        if top_down:
+            get_in_edges = lambda key: sorted([key for key in self.nodes[key].parents])
+            get_out_edges = lambda key: sorted([key for key in self.nodes[key].children])
+        else:
+            get_in_edges = lambda key: sorted([key for key in self.nodes[key].children])
+            get_out_edges = lambda key: sorted([key for key in self.nodes[key].parents])
+
+        # Create a representation of the DAG we can consume as we traverse
+        entire_graph = {
+            key: (key, get_in_edges(key), get_out_edges(key)) for key, _ in self.nodes.items()
+        }
+
+        # Populate a queue with all the nodes that have no incoming edges
+        nodes: Queue = Queue()
+        for _, node in entire_graph.items():
+            _, in_edges, _ = node
+            if len(in_edges) == 0:
+                nodes.put((0, node))
+
+        while not nodes.empty():
+            # Remove the next node, n, from the queue and yield it
+            depth, (n_key, _, n_out_edges) = nodes.get()
+            yield (depth, (n_key, self.nodes[n_key]))
+
+            # for each node, m, with an edge, e, from n to m
+            for m in n_out_edges:
+                _, m_in_edges, _ = entire_graph[m]
+                # remove edge e from the graph
+                m_in_edges.remove(n_key)
```

Comment:
You don't have to remove an edge like the text books suggests, you only have to decrement in edge count.